002-tidyr-stringr.Rmd

```diff
@@ -0,0 +1,133 @@
+---
+output: github_document
+---
+
+# Tidyup 2: tidyr \<-> stringr alignment
+
+**Champion**: Hadley Wickham  
+**Co-champion**: Mine Çetinkaya-Rundel  
+**Status**: Draft
+
+## Abstract
+
+tidyr provides a handful functions like `separate()` and `extract()` that perform string manipulation.
+Unfortunately, these functions are only weakly connected to their stringr equivalents, making them harder to learn and remember.
+This tidyup proposes a new cohesive family of string manipulation functions to tackle common problems that arise during data tidying.
+The existing functions will remain so existing code is unaffected, but will be superseded to steer folks towards a more consistent API.
+
+```{r setup}
+library(tidyr)
+library(stringr)
+```
+
+## Motivation
+
+stringr and tidyr provide surprisingly different interfaces for string manipulation.
+This means that common tasks require integrating two distinct syntaxes and learning more about stringr doesn't improve your use of tidyr.
+
+The root cause of the problem is that stringr was designed before I understood the importance of data frames, so many stringr functions return matrices (or lists of matrices), which makes them hard to use with other tidyverse packages.
+No stringr function returns a data frame, which is particularly painful given that `summarise()` and `mutate()` can now use data frame results to add multiple columns.
+
+To overcome this weakness, tidyr provides a few functions that work directly with data frames.
+These take a data frame and column specification, and return a modified data frame, typically with new columns specified by the call.
+Unfortunately, the tidyr functions have little relationship to their stringr equivalents:
+
+-   `extract()` is similar to `str_match()`
+-   `separate()` with regex is similar to `str_split_fixed()`
+-   `separate()` with numeric columns has no equivalent in stringr.
+-   `separate_rows()` is equivalent to `str_split()`
+
+Additionally, tidyr uses base R regular expressions, not stringr.
+There are relatively few difference between regexp engines, but you need a very different syntax to control options (e.g. ignoring case).
+While it's important to keep tidyr dependencies lightweight (and installing stringr can be a challenge in some environments), it would be nice to use it if available, following the footsteps laid by [Radix Ordering in](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md) [`dplyr::arrange()`](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md).
```

Comment:
"can be a challenge in some environments due to __?__" (I'm not sure what goes in the blank but would be nice to say that here, if possible to state concisely.)